library/core/qsl_base/src/main/java/org/quiltmc/qsl/base/impl/event/EventRegistry.java

```diff
@@ -0,0 +1,113 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.base.impl.event;
+
+import java.util.List;
+import java.util.Map;
+
+import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
+import net.fabricmc.loader.api.FabricLoader;
+import org.jetbrains.annotations.ApiStatus;
+
+import net.minecraft.util.Identifier;
+
+import org.quiltmc.qsl.base.api.event.Event;
+import org.quiltmc.qsl.base.api.event.EventAwareListener;
+import org.quiltmc.qsl.base.api.event.ListenerPhase;
+import org.quiltmc.qsl.base.api.event.client.ClientEventAwareListener;
+import org.quiltmc.qsl.base.api.event.server.DedicatedServerEventAwareListener;
+
+@ApiStatus.Internal
+public final class EventRegistry {
+	private EventRegistry() {
+		throw new UnsupportedOperationException("EventRegistry only contains static definitions.");
+	}
+
+	@SuppressWarnings({"unchecked", "rawtypes"})
+	public static void listenAll(Object listener, Event<?>... events) {
+		var listenedPhases = getListenedPhases(listener.getClass());
+
+		for (var event : events) {
+			if (!event.getType().isAssignableFrom(listener.getClass())) {
+				throw new IllegalArgumentException("Given object " + listener + " is not a listener of event " + event);
+			}
+
+			var phase = listenedPhases.getOrDefault(event.getType(), Event.DEFAULT_PHASE);
+			((Event) event).register(phase, listener);
+		}
+	}
+
+	private static Map<Class<?>, Identifier> getListenedPhases(Class<?> listenerClass) {
+		var map = new Object2ObjectOpenHashMap<Class<?>, Identifier>();
+
+		for (var annotation : listenerClass.getAnnotations()) {
+			if (annotation instanceof ListenerPhase phase) {
+				map.put(phase.callbackTarget(), new Identifier(phase.namespace(), phase.path()));
+			}
+		}
+
+		return map;
+	}
+
+	@SuppressWarnings({"unchecked", "rawtypes"})
+	public static <T> void register(Event<T> event, Class<? super T> type) {
+		for (var target : EventSideTarget.VALUES) {
+			// Search if the callback qualifies is unique to this event.
+			if (target.listenerClass().isAssignableFrom(type)) {
+				List<?> entrypoints = FabricLoader.getInstance().getEntrypoints(target.entrypointKey(), target.listenerClass());
+
+				// Search for matching entrypoint.
+				for (Object entrypoint : entrypoints) {
+					var listenedPhases = getListenedPhases(entrypoint.getClass());
+
+					// Searching if the given entrypoint is a listener of the event being registered.
+					if (type.isAssignableFrom(entrypoint.getClass())) {
```

Comment:
Is this enough is one question I would ask. If an event callback interface has a generic this may be wrong?

Reply:
More specifically we need tests to ensure we don't let undefined behavior in here.

Reply:
If there is a generic, it implies there can be multiple instances of an event for the given callback, so for this register method this is right, since it won't extend EventAwareListener.

Tho for `listenAll` it mays be wrong then, since it uses the event array, I'm not exactly sure how I can compare it?